core/src/main/scala/kafka/server/KafkaConfig.scala

```diff
@@ -1385,10 +1385,22 @@ object KafkaConfig {
     }
     if (maybeSensitive) Password.HIDDEN else value
   }
+
+  def populateSynonyms(input: util.Map[_, _]): util.Map[Any, Any] = {
+    val output = new util.HashMap[Any, Any](input)
+    val brokerId = output.get(KafkaConfig.BrokerIdProp)
+    val nodeId = output.get(KafkaConfig.NodeIdProp)
+    if (brokerId == null && nodeId != null) {
+      output.put(KafkaConfig.BrokerIdProp, nodeId)
+    } else if (brokerId != null && nodeId == null) {
+      output.put(KafkaConfig.NodeIdProp, brokerId)
+    }
+    output
+  }
 }
 
 class KafkaConfig(val props: java.util.Map[_, _], doLog: Boolean, dynamicConfigOverride: Option[DynamicBrokerConfig])
-  extends AbstractConfig(KafkaConfig.configDef, props, doLog) with Logging {
+  extends AbstractConfig(KafkaConfig.configDef, KafkaConfig.populateSynonyms(props), doLog) with Logging {
```

Comment:
It seems a little bit odd that we populate synonyms only in the reference that we're passing to `AbstractConfig`. The field `KafkaConfig.props` could still be accessed directly (maybe it should be private?). Would it make sense to move this to a factory method? For example:

```scala
object KafkaConfig {
  def apply(props: java.util.Map[_, _], doLog: Boolean, dynamicConfigOverride: Option[DynamicBrokerConfig]): KafkaConfig = {
    new KafkaConfig(populateSynonyms(props), doLog, dynamicConfigOverride)
  }
}
```

Reply:
Yes, good find. This is a case where Scala's behavior is kind of annoying. I wish there was a way to opt-out of the auto-initialization.

Anyway, I made the primary constructor private, and put a call to `KafkaConfig#populateSynonyms` in all the (publicly visible) secondary constructors. This should fix it...